lib/resty/iprepd.lua

```diff
@@ -0,0 +1,76 @@
+local cjson = require('cjson')
+local lrucache = require('resty.lrucache')
+local http = require('resty.http')
+
+function fatal_error(error_msg)
+  ngx.log(ngx.ERR, error_msg)
+  os.exit(1)
+end
+
+local _M = {}
+local mt = { __index = _M }
+
+function _M.new(options)
+  local iprepd_url = options.url or 'http://localhost:8080/'
+  local cache_ttl = options.cache_ttl or 30
+  local iprepd_threshold = options.threshold
+  if not iprepd_threshold then
+    fatal_error('Need to pass in a threshold')
+  end
+  local iprepd_api_key = options.api_key
+  if not iprepd_api_key then
+    fatal_error('Need to pass in an api_key')
+  end
+
+  -- TODO: Make configurable?
+  local cache, err = lrucache.new(200)
+  if not cache then
+    fatal_error('failed to create the cache: ' .. (err or 'unknown'))
+  end
+
+  local self = {
+    url = iprepd_url,
+    threshold = iprepd_threshold,
+    api_key_hdr = {
+      ['Authorization'] = 'APIKey ' .. iprepd_api_key,
+    },
+    cache_ttl = cache_ttl,
+    timeout = options.timeout or 10,
+    cache = cache,
+  }
+  return setmetatable(self, mt)
+end
+
+function _M.check(self, ip)
+  local httpc = http.new()
+  -- set timeout in ms
+  httpc:set_timeout(self.timeout)
+
+  -- Get reputation for ip
+  local resp = self.cache:get(ip)
+  if not resp then
+    resp, err = httpc:request_uri(self.url .. '/' .. ip, {
+      method  = "GET",
+      headers = self.api_key_hdr,
+    })
+    if err then
+      ngx.log(ngx.ERR, 'Error with request to iprepd: ' .. err)
+    else
+      self.cache:set(ngx.var.remote_addr, resp, self.cache_ttl)
+    end
+  end
+
+  -- If the IP was found
+  if resp and resp.status == 200 then
+    local resp_body = cjson.decode(resp.body)
+
+    -- check reputation against threshold
+    if resp_body['reputation'] >= self.threshold then
```

Comment:
I think this should be <= here, the reputation score for addresses lower over time.